ngraph/core/reference/src/runtime/reference/tile.cpp

```diff
@@ -14,34 +14,119 @@
 
 using namespace ngraph;
 
-void runtime::reference::tile(
-    const char* arg, char* out, const Shape& in_shape, const Shape& out_shape, size_t elem_size)
+namespace
+{
+    struct Counter
+    {
+        Counter(const Shape& shape)
+            : m_shape(shape)
+        {
+            m_indices.resize(m_shape.size() - 1, 0);
+            m_axis = m_indices.size();
+        }
+
+        size_t get_axis() { return m_axis; }
+        bool get_run() { return m_run; }
+        bool is_axis_upper_bound()
+        {
+            if (m_axis-- == 0)
+            {
+                m_run = false;
+                return false;
+            }
+
+            if (++m_indices[m_axis] != m_shape[m_axis])
+            {
+                m_axis = m_indices.size();
+                return false;
+            }
+
+            m_indices[m_axis] = 0;
+            return true;
+        }
+
+    private:
+        std::vector<int64_t> m_indices;
+        size_t m_axis;
+        bool m_run{true};
+        Shape m_shape;
+    };
+
+    /// \brief For each axis calculates the product of inner axes
+    /// If dims has shape (2, 3, 4) then for 2 (first axis) the inner axes would be (3, 4)
+    /// and for 3 (second axis) it would be (4)
+    /// If dims has shape(2, 3, 4) then the output vector would be (3 * 4, 4, 1)
+    /// The outermost axis is not used. For innermost axis it is always 1.
+    /// \param[in] dims Shape of the output
+    ///
+    /// \return Vector containing calculated values for each axis.
+    std::vector<int64_t> create_pitches(const Shape& dims)
+    {
+        std::vector<int64_t> pitch;
+        pitch.resize(dims.size());
+        std::partial_sum(dims.rbegin(), dims.rend() - 1, pitch.begin(), std::multiplies<int64_t>());
+        std::reverse(pitch.begin(), pitch.end() - 1);
+
+        pitch.back() = 1;
+        return pitch;
+    }
+}
+
+void runtime::reference::tile(const char* arg,
+                              char* out,
+                              const Shape& in_shape,
+                              const Shape& out_shape,
+                              const size_t elem_size,
+                              const std::vector<int64_t>& repeats)
 {
     Shape in_shape_expanded(in_shape);
     in_shape_expanded.insert(in_shape_expanded.begin(), out_shape.size() - in_shape.size(), 1);
-    CoordinateTransform input_transform(in_shape_expanded);
-    CoordinateTransform output_transform(out_shape);
+    size_t block_size = 0;
+    int64_t num_repeats = 0;
+    std::vector<int64_t> indices(in_shape_expanded.size() - 1, 0);
+    size_t axis(indices.size());
+    const int input_rank = in_shape_expanded.size();
+    const int64_t last_dim = in_shape_expanded[input_rank - 1];
+    const std::vector<int64_t> pitches = create_pitches(out_shape);
+    const char* copy = nullptr;
+    Counter counter(in_shape_expanded);
 
-    for (const Coordinate& output_coord : output_transform)
+    while (counter.get_run())
```

Comment:
I propose to change this loop with run function. I don't think that we need a special counter only for two conditionals.

This conditional you can change to something like: `axis < in_shape_expanded.size()`.